docs/spec/connection-handshake/L1_2.md

```diff
@@ -0,0 +1,416 @@
+  counterpartyPrefix                CommitmentPrefix
+  clientIdentifier                  Identifier
+  counterpartyClientIdentifier      Identifier
+  version                           string | []string
+}
+```
+
+
+```golang
+type ConnectionEnd interface {
+    state      ConnectionState
+    initiator  ConnectionParty
+    responder  ConnectionParty
+    role       ConnectionRole   // either 'initiator' or 'responder'
+}
+```
+
+Hereafter, when we use the term connection, we refer to the above definition of this type.
+
+
+We now restate the connection handshake problem in a slightly more precise (or lower level) formulation:
+A connection handshake enables two processes to agree on a consistent view of their state, to verify each other's state, and to allocate a new connection.
+
+###### Local state & variables
+The local state consists of the following:
+
+- An object (variable) of type `ConnectionEnd`, labeled `c`, initialized to `null` at each of processes `I` and `R`.
+- `State` `si` is initialized to `null` at process `I`.
+- `State` `sr` is initialized to `null` at process `R`.
+
+
+###### Inputs
+There are three functions that process `I` and `R` have access to:
+
+1. Function `unique(ConnectionEnd c)` takes as input a `ConnectionEnd` `c` and returns true if the input value `c` is unique: if there was no connection established so far with the same `ConnectionEnd`.
+
+2. Function `verify(State s, Process x)` returns true or false.
+    - Process `I` invokes this as follows: `verify(si, R)`, where `si` is the representation of the state of process `R` according to process `I`. This function return trues iff `si` passes verification, and false otherwise.
+    - The invokation and properties at process `R` is analogous.
+
+3. Function `consistent(State s, Process x)` return true or false. This is similar in purpose and function with `verify`, except is check the consistency of the state `s` as stored locally (at the invoker) with the same state remotely at process `x`.
+
+Furthermore, process `I` receives as input the following:
+
+- `State` `si` and `State` `sr`, representing the state of
+
+# Stable guarantees start here
+
+###### Guarantees
+We refine the safety guarantee **[CH/Seq/1/Agreement]** (defined in the [sequential problem statement](#sequential-problem-statement)) by splitting this into four complementary safety properties:
+
+- **[CH/Proto/1/Validity]**
+If a module establishes a `Connection` `c`, then `c` is a _valid_ connection; validity is defined by an application-level predicate.
+
+- **[CH/Proto/2/Uniqueness]**
+If a module establishes a connection `c`, no handshake has previously finished for `c`; uniqueness is defined by an application-level predicate.
+
+- **[CH/Proto/6/Termination]**
+Eventually, the connection handshake terminates, i.e., both parties mark their respective `ConnectionEnd` as `open`.
+
+- **[CH/Proto/7/Integrity]**
+If either Alice or Bob establishes a `Connection` `c`, then Alice proposed `c`.
```

Comment:
That's a very good point. I only understood later that any party can be the proposer in a connection. This is going to be amended to the following version:

> If any two modules establish a connection `c`, then either one module or the other or both modules proposed `c`.

It might be redundant to say "or both" but keeping it like that for the moment.

Reply:
I think "or both" should stay there, as it captures that *at least one* of the modules proposed `c`. You could also consider to formulate it as the contrapositive:
> If neither Alice nor Bob proposed a connection, then no connection between Alice and Bob is established.